src/main/java/net/minecraftforge/server/ServerLifecycleHooks.java

```diff
@@ -147,9 +151,22 @@ public static boolean handleServerLogin(final ClientIntentionPacket packet, fina
             final ConnectionType connectionType = ConnectionType.forVersionFlag(packet.getFMLVersion());
             final int versionNumber = connectionType.getFMLVersionNumber(packet.getFMLVersion());
 
-            if (connectionType == ConnectionType.MODDED && versionNumber != NetworkConstants.FMLNETVERSION) {
-                rejectConnection(manager, connectionType, "This modded server is not impl compatible with your modded client. Please verify your Forge version closely matches the server. Got net version "+ versionNumber + " this server is net version "+ NetworkConstants.FMLNETVERSION);
-                return false;
+            if (connectionType == ConnectionType.MODDED)
+            {
+                // Allow clients with incorrect netcode version to connect to netversion 3 servers,
+                // if and only if client is netversion 2 and server has no syncable non-vanilla datapack registries.
+                // TODO 1.19: Remove netcode backwards-compatability in 1.19, as there will be no clients on netversion 2 in 1.19. 
+                Set<ResourceKey<? extends Registry<?>>> customDatapackRegistries = DataPackRegistriesHooks.getSyncedCustomRegistries();
+                if (versionNumber == 2) {
+                    if (!customDatapackRegistries.isEmpty()) {
```

Comment:
Formatting: braces on newlines.
```suggestion
                if (versionNumber == 2)
                {
                    if (!customDatapackRegistries.isEmpty())
                    {
```

Reply:
Changed in next push